### DISCUSSION TITLE
Asynchronous Component Run Loop

### DISCUSSION THREAD

**Discussion by SterlingPeet** [nasa/fprime#1779] [Q&A] [answered]
I took a cursory glance at the Users Guide, but I was unable to find a good description of how to approach this problem.

If I have an active component that is doing an asynchronous work, what is the expected pattern for this?

I realize that many things can be formulated to run in some chunked work format (e.g.: saving file data to an SD card), but there are a select few things that don't fit this format very well.  Either the goal might be to eek out a little more performance from a low priority task, or its something like listening for data coming in on a comms interface and lower latency is preferred.

Should I be looking at the Active Component Base (override ``loop()``), or somewhere else?

https://github.com/nasa/fprime/blob/91c578ee37c01ba92e556dbca0287d82f9f18b95/Fw/Comp/ActiveComponentBase.cpp#L133-L149

**Comment by LeStarch** [comment, marked as answer]
Typically we handle this pattern by creating an instance of `Os::Task` to perform the work and attaching that to a component.  The active component loop is designed as a message handling/dispatch mechanism and changing that would break the component behavior.

A good set of examples are Drv/LinuxUartDriver. It starts a read task and passes it a reference to the component.  Thus it has a thread for work, but it isn't abusing the active component thread.

Some other notes on this strategy:
1. Component need not be `active` in this case.  It could be passive + a member task/thread
2. When executing in a baremetal context, thread method patterns are important.

Does this answer your question or have I missunderstood?

**Reply by SterlingPeet** [reply]
This is the answer I was looking for.

As a followup, Is it reasonable to replace ``loop()`` with a version that unloads the port queues [if they have waiting calls] and then does a smallish chunk of work [if it has work waiting to be done]?

Or is this better served by an internal port call and avoiding touching loop?

Or alternatively, does it make sense to override loop if the component produces port calls but doesn’t receive any?

I might be getting off into the hypothetical weeds here.  Mostly I have been exploring ideas that might have some odd execution patterns, like a buffer scrubbing component that cycles through a pool custom FEC protected buffers and checks them for bit flips.
